risc0/zkvm/src/host/server/exec/executor.rs

```diff
@@ -294,34 +297,40 @@ impl<'a> ExecutorImpl<'a> {
                             .context("Too many segments to fit in u32")?,
                         cycles,
                     );
-                    let segment_ref = callback(segment)?;
-                    self.segments.push(segment_ref);
```

Comment:
Why not do this for all arms? I would expect the callback to happen for all cases (including Halted and Fault)

Reply:
I think I see now that there is special processing for the final segment. I guess this is fine but makes me feel like a deeper refactor might be needed in the future.

Reply:
I agree. It's definitely not the cleanest code I've ever written 😅